protocol/src/main/java/org/sdo/pri/OnDieSignatureValidator.java

```diff
@@ -0,0 +1,185 @@
+   * @throws Exception when error.
+   */
+  public static boolean validateWithoutRevocations(
+          String signedData,
+          byte[] signature,
+          PublicKey publicKey) {
+
+    byte[] adjSignature = new byte[0];
+    try {
+      // first 36 bytes are always the taskInfo
+      byte[] taskInfo = Arrays.copyOfRange(signature, 0, 36);
+
+      // adjust the signed data
+      // data-to-verify format is: [ task-info | nonce (optional) | data ]
+      // First 36 bytes of signature is the taskinfo. This value must be prepended
+      // to the signedData
+      ByteArrayOutputStream adjSignedData = new ByteArrayOutputStream();
+      adjSignedData.write(taskInfo);
+      adjSignedData.write(signedData.getBytes());
+
+      adjSignature = convertSignature(signature, taskInfo);
+
+      Signature sig = Signature.getInstance("SHA384withECDSA", BouncyCastleLoader.load());
+      sig.initVerify(publicKey);
+      sig.update(adjSignedData.toByteArray());
+      return sig.verify(adjSignature);
+    } catch (Exception ex) {
+      return false;
+    }
+  }
+
+  /**
+   * Performs a validation of the given signature with the public key
+   * extracted from the given cert chain.
+   *
+   * @param signedData signedData
+   * @param signature signature
+   * @param certPath certificate path (dc value from ov)
+   * @param onDieCache cache that contains the CRLs
+   * @return boolean indicating if signature is valid.
+   * @throws Exception when error.
+   */
+  public static boolean validateWithRevocations(
+          String signedData,
+          byte[] signature,
+          CertPath certPath,
+          OnDieCache onDieCache) {
+
+    // Check revocations first.
+    List<Certificate> certificateList = (List<Certificate>) certPath.getCertificates();
+
+    // Check revocations first.
+    if (!checkRevocations(certificateList, onDieCache)) {
+      return false;
+    }
+
+    // Now validate the signature
+    // Public key comes from the cert chain in the voucher.
+    return validateWithoutRevocations(signedData, signature, certificateList.get(0).getPublicKey());
+  }
+
+
+  private static byte[] convertSignature(byte[] signature, byte[] taskInfo) throws Exception {
+    // 0x30 b1 0x02 b2 (vr) 0x02 b3 (vs)
+    byte[] rvalue = Arrays.copyOfRange(signature, taskInfo.length, taskInfo.length + 48);
+    byte[] svalue = Arrays.copyOfRange(signature, taskInfo.length + 48, taskInfo.length + 96);
+
+    boolean appendZeroToR = false;
+    boolean appendZeroToS = false;
+    if ((rvalue[0] & 0x80) != 0) {
+      appendZeroToR = true;
+    }
+    if ((svalue[0] & 0x80) != 0) {
+      appendZeroToS = true;
+    }
+
+    ByteArrayOutputStream adjSignature = new ByteArrayOutputStream();
+    adjSignature.write(0x30);
+    // total length of remaining bytes
+    adjSignature.write(4 + (appendZeroToR ? 49 : 48) + (appendZeroToS ? 49 : 48));
+    adjSignature.write(0x02);
+    // R value
+    if (appendZeroToR) {
+      adjSignature.write(49);
```

Comment:
This code is duplicated on lines 136-141.  Should it be pulled into a function to prevent maintenance errors?